java/core/src/test/org/apache/orc/impl/TestRecordReaderUtils.java

```diff
@@ -140,6 +146,44 @@ public void testExtraBytesReadWithinThreshold() {
     assertEquals(chunkReader.getReadBytes(), chunkReader.getFrom().getData().array().length);
   }
 
+  @Test
+  public void testZeroCopyReadAndRelease() throws IOException {
+    int blockSize = 4096;
+    ByteBuffer hdfsBlockMMapBuffer = makeByteBuffer(blockSize, 0);
+    int blockStartPosition = 4096;
+    MockDFSDataInputStream dis = new MockDFSDataInputStream(hdfsBlockMMapBuffer, blockStartPosition);
+    FSDataInputStream fis = new FSDataInputStream(dis);
+    RecordReaderUtils.ByteBufferAllocatorPool pool = new RecordReaderUtils.ByteBufferAllocatorPool();
+    HadoopShims.ZeroCopyReaderShim zrc = RecordReaderUtils.createZeroCopyShim(fis, null, pool);
+    BufferChunkList rangeList = new TestOrcLargeStripe.RangeBuilder()
+            .range(5000, 1000)
+            .range(6000, 1000)
+            .range(7000, 500).build();
+    RecordReaderUtils.zeroCopyReadRanges(fis, zrc, rangeList.get(0), rangeList.get(2), false);
+
+    assertArrayEquals(Arrays.copyOfRange(hdfsBlockMMapBuffer.array(), 5000 - blockStartPosition, 5000 - blockStartPosition + 1000), byteBufferToArray(rangeList.get(0).getData()));
+    assertArrayEquals(Arrays.copyOfRange(hdfsBlockMMapBuffer.array(), 6000 - blockStartPosition, 6000 - blockStartPosition + 1000), byteBufferToArray(rangeList.get(1).getData()));
+    assertArrayEquals(Arrays.copyOfRange(hdfsBlockMMapBuffer.array(), 7000 - blockStartPosition, 7000 - blockStartPosition + 500), byteBufferToArray(rangeList.get(2).getData()));
+
+    assertThrowsExactly(IllegalArgumentException.class, new Executable() {
```

Comment:
This `IllegalArgumentException` is our mock code. May I ask what happens previously in old ORC?

Reply:
The previously old ORC code will throw a `IllegalArgumentException` when releaseBuffer(), this assert is used to prove it.

Reply:
`IllegalArgumentException ` is from hadoop DFSInputStream https://github.com/apache/hadoop/blob/trunk/hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DFSInputStream.java#L1921